MikroOrm minimum Node version (util/types)

I know the mikro orm adapter is new. (Thanks so much for it btw).

Just started trying it out and got an error 

```
Error: Cannot find module 'util/types'
```

I was on v14.16.1 upgraded to v16.13.1 and it worked. Didn't look into it exactly when node split the modules apart but might be good to just give warning for it. Also just wanted to post it here in case anyone else tries it.

--- a/docs/docs/adapters/mikro-orm.md
+++ b/docs/docs/adapters/mikro-orm.md
@@ -6,7 +6,7 @@ title: MikroORM
 To use this Adapter, you need to install Mikro ORM, the driver that suits your database, and the separate `@next-auth/mikro-orm-adapter` package:
 
 ```bash npm2yarn
-npm install next-auth @next-auth/mikro-orm-adapter @mikro-orm/core@latest @mikro-orm/[YOUR DRIVER]@latest
+npm install next-auth @next-auth/mikro-orm-adapter @mikro-orm/core @mikro-orm/[YOUR DRIVER]
 ```
 
 Configure NextAuth.js to use the MikroORM Adapter:

--- a/docs/docs/adapters/mikro-orm.md
+++ b/docs/docs/adapters/mikro-orm.md
@@ -6,7 +6,7 @@ title: MikroORM
 To use this Adapter, you need to install Mikro ORM, the driver that suits your database, and the separate `@next-auth/mikro-orm-adapter` package:
 
 ```bash npm2yarn
-npm install next-auth @next-auth/mikro-orm-adapter @mikro-orm/core@next @mikro-orm/[YOUR DRIVER]@next
+npm install next-auth @next-auth/mikro-orm-adapter @mikro-orm/core@latest @mikro-orm/[YOUR DRIVER]@latest
 ```
 
 Configure NextAuth.js to use the MikroORM Adapter:
@@ -31,7 +31,7 @@ export default NextAuth({
 
 ### Passing custom entities
 
-The MikroORM adapter ships with its own set of entities. If you'd like to override them, you can optionally pass them to the adapter.
+The MikroORM adapter ships with its own set of entities. If you'd like to extend them, you can optionally pass them to the adapter.
 
 > This schema is adapted for use in MikroORM and based upon our main [schema](/adapters/models)
 

--- a/packages/adapter-mikro-orm/README.md
+++ b/packages/adapter-mikro-orm/README.md
@@ -35,7 +35,6 @@ This is the MikroORM Adapter for [`next-auth`](https://next-auth.js.org). This p
    export default NextAuth({
       // https://next-auth.js.org/configuration/providers
       providers: [],
-      // optionally pass extended models as { entities: { } }
       adapter: MikroOrmAdapter({
          dbName: "./db.sqlite",
          type: "sqlite",

--- a/packages/adapter-mikro-orm/README.md
+++ b/packages/adapter-mikro-orm/README.md
@@ -41,6 +41,8 @@ This is the MikroORM Adapter for [`next-auth`](https://next-auth.js.org). This p
          type: "sqlite",
          debug: process.env.DEBUG === "true" || process.env.DEBUG?.includes("db"),
          ...
+      }, {
+         // pass extended models as { entities: { } } if needed
       }),
       ...
    });

--- a/packages/adapter-mikro-orm/package.json
+++ b/packages/adapter-mikro-orm/package.json
@@ -32,12 +32,12 @@
     "dist"
   ],
   "peerDependencies": {
-    "@mikro-orm/core": "^5.0.0",
+    "@mikro-orm/core": "^5.0.2",
     "next-auth": "^4.0.1"
   },
   "devDependencies": {
-    "@mikro-orm/core": "^5.0.0",
-    "@mikro-orm/sqlite": "^5.0.0"
+    "@mikro-orm/core": "^5.0.2",
+    "@mikro-orm/sqlite": "^5.0.2"
   },
   "jest": {
     "preset": "adapter-test/jest"

--- a/packages/adapter-mikro-orm/src/index.ts
+++ b/packages/adapter-mikro-orm/src/index.ts
@@ -1,55 +1,66 @@
-import type { Options } from "@mikro-orm/core"
+import type {
+  AnyEntity,
+  EntityClass,
+  EntityClassGroup,
+  EntitySchema,
+  Options as ORMOptions,
+} from "@mikro-orm/core"
 import { MikroORM, wrap } from "@mikro-orm/core"
 import * as defaultEntities from "./entities"
 
 import type { Adapter } from "next-auth/adapters"
-import { isPromise } from "util/types"
 
 export * as defaultEntities from "./entities"
 
 /**
- * The MikroORM adapter accepts a MikroORM instance or configuration and returns a NextAuth adapter.
- * @param ormConnection can either be an instance promise or a MikroORM connection configuration (https://mikro-orm.io/docs/next/configuration#driver)
- * @param options entities in the options object will be passed to the MikroORM init function as entities
+ * The MikroORM adapter accepts a MikroORM configuration and returns a NextAuth adapter.
+ * @param ormConnection a MikroORM connection configuration (https://mikro-orm.io/docs/next/configuration#driver)
+ * @param options entities in the options object will be passed to the MikroORM init function as entities. Has to be provided if overridden!
  * @returns
  */
 export function MikroOrmAdapter(
-  ormConnection: Promise<MikroORM> | Options,
+  ormOptions: ORMOptions,
   options?: {
     entities?: Partial<typeof defaultEntities>
   }
 ): Adapter {
-  const {
-    User: UserModel,
-    Account: AccountModel,
-    Session: SessionModel,
-    VerificationToken: VerificationTokenModel,
-  } = { ...defaultEntities, ...options?.entities }
+  const UserModel = options?.entities?.User ?? defaultEntities.User
+  const AccountModel = options?.entities?.Account ?? defaultEntities.Account
+  const SessionModel = options?.entities?.Session ?? defaultEntities.Session
+  const VerificationTokenModel =
+    options?.entities?.VerificationToken ?? defaultEntities.VerificationToken
+
+  let _orm: MikroORM
 
   const getEM = async () => {
-    if (!isPromise(ormConnection)) {
-      if (typeof ormConnection.entities === "string")
-        throw new Error("You have to pass class entities to MikroORM.init")
-
+    if (!_orm) {
       // filter out default entities from the passed entities
-      const connectionEntities = ormConnection.entities?.filter((e) => {
-        if (typeof e !== "string" && "name" in e && typeof e.name === "string")
-          return !["User", "Account", "Session", "VerificationToken"].includes(
-            e.name
+      const optionsEntities =
+        ormOptions.entities?.filter((e) => {
+          if (
+            typeof e !== "string" &&
+            "name" in e &&
+            typeof e.name === "string"
           )
-        return true
-      })
+            return ![
+              "User",
+              "Account",
+              "Session",
+              "VerificationToken",
+            ].includes(e.name)
+          return true
+        }) ?? []
       // add the (un-)enhanced entities to the connection
-      ormConnection.entities = [
-        ...(connectionEntities ?? []),
+      ormOptions.entities = [
+        ...optionsEntities,
         UserModel,
         AccountModel,
         SessionModel,
         VerificationTokenModel,
       ]
-      ormConnection = MikroORM.init(ormConnection)
+      _orm = await MikroORM.init(ormOptions)
     }
-    return await ormConnection.then((orm) => orm.em.fork())
+    return _orm.em.fork()
   }
 
   return {

--- a/packages/adapter-mikro-orm/src/index.ts
+++ b/packages/adapter-mikro-orm/src/index.ts
@@ -1,8 +1,10 @@
 import type {
   AnyEntity,
+  Connection,
   EntityClass,
   EntityClassGroup,
   EntitySchema,
+  IDatabaseDriver,
   Options as ORMOptions,
 } from "@mikro-orm/core"
 import { MikroORM, wrap } from "@mikro-orm/core"
@@ -18,8 +20,8 @@ export * as defaultEntities from "./entities"
  * @param options entities in the options object will be passed to the MikroORM init function as entities. Has to be provided if overridden!
  * @returns
  */
-export function MikroOrmAdapter(
-  ormOptions: ORMOptions,
+export function MikroOrmAdapter<D extends IDatabaseDriver<Connection> = IDatabaseDriver<Connection>>(
+  ormOptions: ORMOptions<D>,
   options?: {
     entities?: Partial<typeof defaultEntities>
   }

--- a/packages/adapter-mikro-orm/tests/index.test.ts
+++ b/packages/adapter-mikro-orm/tests/index.test.ts
@@ -1,4 +1,6 @@
 import type { Options } from "@mikro-orm/core"
+import type { SqliteDriver } from "@mikro-orm/sqlite"
+
 import { MikroORM, wrap } from "@mikro-orm/core"
 import { runBasicTests } from "adapter-test"
 import { MikroOrmAdapter, defaultEntities } from "../src"
@@ -14,7 +16,7 @@ const entities = [
   VeryImportantEntity,
 ]
 
-const config: Options = {
+const config: Options<SqliteDriver> = {
   dbName: "./db.sqlite",
   type: "sqlite",
   entities,

--- a/packages/adapter-mikro-orm/tests/index.test.ts
+++ b/packages/adapter-mikro-orm/tests/index.test.ts
@@ -1,3 +1,4 @@
+import type { Options } from "@mikro-orm/core"
 import { MikroORM, wrap } from "@mikro-orm/core"
 import { runBasicTests } from "adapter-test"
 import { MikroOrmAdapter, defaultEntities } from "../src"
@@ -13,34 +14,22 @@ const entities = [
   VeryImportantEntity,
 ]
 
+const config: Options = {
+  dbName: "./db.sqlite",
+  type: "sqlite",
+  entities,
+  debug: process.env.DEBUG === "true" || process.env.DEBUG?.includes("db"),
+}
+
 async function getORM() {
   if (_init) return _init
 
-  _init = await MikroORM.init({
-    dbName: "./db.sqlite",
-    type: "sqlite",
-    entities,
-    debug: process.env.DEBUG === "true" || process.env.DEBUG?.includes("db"),
-  })
+  _init = await MikroORM.init(config)
   return _init
 }
 
 runBasicTests({
-  adapter: MikroOrmAdapter(
-    {
-      dbName: "./db.sqlite",
-      type: "sqlite",
-      entities: [
-        defaultEntities.User,
-        defaultEntities.Account,
-        defaultEntities.Session,
-        defaultEntities.VerificationToken,
-        VeryImportantEntity,
-      ],
-      debug: process.env.DEBUG === "true" || process.env.DEBUG?.includes("db"),
-    },
-    { entities: { User } }
-  ),
+  adapter: MikroOrmAdapter(config, { entities: { User } }),
   db: {
     async connect() {
       const orm = await getORM()

